### PR TITLE
Add Validation for date_left Column in Residents Table

### LINKED ISSUE #100: Add Check for date_left in Residents
Currently, its possible to have a date_left value that occurs prior to a date_joined date. There should be a check to make sure this doesn't occur (ideally in the APIs) and throw some error if it does

### LINKED ISSUE #104: GET API in Residents Needs Requires Auth By Role
```@blueprint.route("/", methods=["GET"], strict_slashes=False)
def get_resident():
```

Needs to be updated to:
```
@blueprint.route("/", methods=["GET"], strict_slashes=False)
@require_authorization_by_role(<ALL ROLES>)
def get_resident():
```

### PR DESCRIPTION
**Pull request by safchow**
## GitHub Issue link

#100 

## Implementation description
* Added table constraint for `date_left` in the model
* Added validation in the routes layer of the `POST` and `PUT` APIs

### Database Tests
1. Verify this insert fails with a constraint error: 
`INSERT INTO residents (initial, room_num, date_joined, date_left, building) VALUES ('sc', '420', '2022-02-27', '2022-02-27', '402');`
2. Verify this insert runs successfully:
`INSERT INTO residents (initial, room_num, date_joined, date_left, building) VALUES ('sc', '420', '2022-02-27', '2022-02-28', '402');`

### API Tests
1. Verify the `POST` API correctly validates date_left > date_joined. Here this JSON body fails:
```
{
    "initial": "sc",
    "room_num": 148,
    "date_joined": "2023-02-27",
    "date_left": "2023-02-26",
    "building": "402"
}
```
2. Change the date_left so that date_left > date_joined. Verify you receive a 201 response.
3. Using the same JSON body, verify the `PUT` API makes the same validation. Verify you get a 400 response when updating an existing resident with an invalid date_left value, and a 201 when it is valid.
 

## What should reviewers focus on?
* beep boop

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR
- [x] If I have made API changes, I have updated the [REST API Docs](https://www.notion.so/uwblueprintexecs/REST-Endpoints-05ce60312bb943439dfda42bb1318536)
- [x] IF I have made changes to the db/models, I have updated the [Data Models Page](https://www.notion.so/uwblueprintexecs/Data-Models-760f8aa06b244eb0842c079ad77987b0)
- [x] I have updated other Docs as needed

### CODE REVIEW

File: backend/app/services/implementations/residents_service.py [hunk 1]
```diff
@@ -81,13 +82,13 @@ def delete_resident(self, resident_id):
             )
         db.session.commit()
 
-    def get_resident(self, resident_id=None):
+    def get_residents(self, resident_id=None):
```
Comment (helenellyx) [inline backend/app/services/implementations/residents_service.py:85]: oopsie daisy, u forgor to change the name in the interface

**Review by helenellyx — CHANGES_REQUESTED** [review]
LOOKS GOOD! the API calls are all working beautifully, but when I use raw sql, it inserts the resident regardless of their date_left. I suspect that the table constraint "check_date_left_valid" is not being properly upgrade by flask 🤔 lmk what you think

**Comment by safchow** [comment]
> LOOKS GOOD! the API calls are all working beautifully, but when I use raw sql, it inserts the resident regardless of their date_left. I suspect that the table constraint "check_date_left_valid" is not being properly upgrade by flask 🤔 lmk what you think

You need to do a fresh docker re-run for the changes to take effect, but its super tedious and you'll have to re-build the containers for whatever you're currently working on. I've tested the raw sql inserts and they threw errors so they should work- we can skip testing them and just test the APIs :D

File: backend/app/services/implementations/residents_service.py (hunk 1 above)
Comment (safchow) [reply backend/app/services/implementations/residents_service.py:85]: I forgor 💀

**Review by helenellyx — APPROVED** [review]
absolutely beasted LGTM

**Review by owen-sellner — APPROVED** [review]
LGTM!!! It passed all the API tests.

**Pull request merged**